pkg/fuzzer/queue/queue.go

```diff
@@ -0,0 +1,375 @@
+		for pq.pos > 0 {
+			newLen := len(pq.queue) - 1
+			pq.queue[newLen] = nil
+			pq.queue = pq.queue[:newLen]
+			pq.pos--
+		}
+	}
+	pq.queue = append(pq.queue, req)
+}
+
+func (pq *PlainQueue) Next() *Request {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	return pq.nextLocked()
+}
+
+func (pq *PlainQueue) tryNext() *Request {
+	if !pq.mu.TryLock() {
+		return nil
+	}
+	defer pq.mu.Unlock()
+	return pq.nextLocked()
+}
+
+func (pq *PlainQueue) nextLocked() *Request {
+	if pq.pos < len(pq.queue) {
+		ret := pq.queue[pq.pos]
+		pq.queue[pq.pos] = nil
+		pq.pos++
+		if pq.stat != nil {
+			pq.stat.Add(-1)
+		}
+		return ret
+	}
+	return nil
+}
+
+// SourceMultiplexer combines several different sources in a particular order.
+// It is NOT thread-safe.
+type SourceMultiplexer struct {
+	sources []Source
+}
+
+func Multiplex(sources ...Source) Source {
+	return &SourceMultiplexer{sources: sources}
+}
+
+func (sm *SourceMultiplexer) Next() *Request {
+	for _, s := range sm.sources {
+		req := s.Next()
+		if req != nil {
+			return req
+		}
+	}
+	return nil
+}
+
+type callback struct {
+	cb func() *Request
+}
+
+// Callback produces a source that calls the callback to serve every Next() request.
+func Callback(cb func() *Request) Source {
+	return &callback{cb}
+}
+
+func (cb *callback) Next() *Request {
+	return cb.cb()
+}
+
+type alternate struct {
+	base Source
+	mu   sync.Mutex
+	rnd  *rand.Rand
+	prob float32
+}
+
+// Alternate returns a nil *Request in `prob` share of requests.
+func Alternate(base Source, rnd *rand.Rand, prob float32) Source {
+	return &alternate{
+		base: base,
+		rnd:  rnd,
+		prob: prob,
+	}
+}
+
+func (a *alternate) Next() *Request {
+	var skip bool
+	if a.mu.TryLock() {
+		skip = a.rnd.Float32() < a.prob
+		a.mu.Unlock()
+	}
+	if skip {
+		return nil
+	}
+	return a.base.Next()
+}
+
+type PriorityQueue struct {
+	mu       *sync.Mutex
+	ops      *priorityQueueOps[*Request]
+	currPrio priority
+}
+
+func Priority() *PriorityQueue {
+	return &PriorityQueue{
+		mu:       &sync.Mutex{},
+		ops:      &priorityQueueOps[*Request]{},
+		currPrio: priority{0},
+	}
+}
+
+func (pq *PriorityQueue) AppendQueue() *PriorityQueue {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	pq.currPrio = pq.currPrio.next()
+	nextPrio := append(priority{}, pq.currPrio...)
+	return &PriorityQueue{
+		// We use the same queue, therefore the same mutex.
+		mu:       pq.mu,
+		ops:      pq.ops,
+		currPrio: append(nextPrio, 0),
+	}
+}
+
+func (pq *PriorityQueue) Submit(req *Request) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	pq.currPrio = pq.currPrio.next()
+	pq.ops.Push(req, pq.currPrio)
+}
+
+func (pq *PriorityQueue) Next() *Request {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	return pq.ops.Pop()
+}
+
+type DynamicMultiplexer struct {
+	mu   sync.Mutex
+	list []Source
+}
+
+func (dm *DynamicMultiplexer) Add(source Source) {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+	dm.list = append(dm.list, source)
+}
+
+func (dm *DynamicMultiplexer) Next() *Request {
+	dm.mu.Lock()
```

Comment:
You use tryLock in a number of places to avoid blocking and synchronization overheads, but with this DynamicMultiplexer we now always lock this global mutex for each request in Exchange handlers, and then lock the global vminfo.Checker queue, just to see that it's empty (after checking is done).
Maybe we could use atomic.Value to store the current Source and replace checker with fuzzer.

Reply:
I've replaced `DynamicMultiplexer` with `DynamicSource` that allows to dynamically control only one Source. That turned out to be enough for our use cases.

Reply:
This looks better. I suspect it may be even simpler/faster. See the other comment in rpc.go. Let's continue there.